app/components/primer/octicon_component.rb

```diff
@@ -27,20 +29,56 @@ class OcticonComponent < Primer::Component
     # @param size [Symbol] <%= one_of(Primer::OcticonComponent::SIZE_MAPPINGS) %>
     # @param system_arguments [Hash] <%= link_to_system_arguments_docs %>
     def initialize(icon_name = nil, icon: nil, size: SIZE_DEFAULT, **system_arguments)
-      @icon = icon_name || icon
-      @system_arguments = system_arguments
+      icon_key = icon_name || icon
+      cache_key = [icon_key, size, system_arguments].to_s
+
+      if (cache_icon = Primer::OcticonComponent::Cache.read(cache_key))
+        @icon, @system_arguments = cache_icon
+      else
+        @system_arguments = system_arguments
+        @system_arguments[:tag] = :svg
+
+        # Filter out classify options to prevent them from becoming invalid html attributes.
+        # Note height and width are both classify options and valid html attributes.
+        octicon_options = {
+          height: SIZE_MAPPINGS[fetch_or_fallback(SIZE_OPTIONS, size, SIZE_DEFAULT)]
+        }.merge(@system_arguments.slice(:height, :width))
 
-      @system_arguments[:class] = Primer::Classify.call(**@system_arguments)[:class]
-      @system_arguments[:height] ||= SIZE_MAPPINGS[size]
+        @icon = Octicons::Octicon.new(icon_key, octicon_options)
 
-      # Filter out classify options to prevent them from becoming invalid html attributes.
-      # Note height and width are both classify options and valid html attributes.
-      octicon_helper_options = @system_arguments.slice(:height, :width)
-      @system_arguments = add_test_selector(@system_arguments).except(*Primer::Classify::VALID_KEYS, :classes).merge(octicon_helper_options)
+        @system_arguments[:classes] = class_names(
+          @icon.options[:class],
+          @system_arguments[:classes]
+        )
+        @system_arguments.merge!(@icon.options.except(:class))
+
+        Primer::OcticonComponent::Cache.set(cache_key, [@icon, @system_arguments])
+      end
     end
 
     def call
-      octicon(@icon, { **@system_arguments })
+      render(Primer::BaseComponent.new(**@system_arguments)) { @icon.path.html_safe } # rubocop:disable Rails/OutputSafety
+    end
+
+    # :nodoc:
+    class Cache
```

Comment:
I wonder if we could pre-populate some icons here, like we do with `Classify::Cache`

Reply:
I don't know if we can make the assumption of what icons we'll be accessing. This also caches based on all system arguments passed to the component.

Reply:
I think it would be good to have simple cases, we have some icons that are used a lot:
e.g.: `x` icon without any arguments